library/x509_crt.c

```diff
@@ -595,6 +595,129 @@ static int x509_get_ext_key_usage( unsigned char **p,
     return( 0 );
 }
 
+/*
+* SubjectKeyIdentifier ::= KeyIdentifier
+*
+* KeyIdentifier ::= OCTET STRING
+*/
+static int x509_get_subject_key_id(unsigned char** p,
+    const unsigned char* end,
+    mbedtls_x509_buf* subject_key_id)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t len = 0u;
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+        MBEDTLS_ASN1_OCTET_STRING)) != 0)
+    {
+        return(ret);
+    }
+    else
+    {
+        subject_key_id->len = len;
+        subject_key_id->tag = MBEDTLS_ASN1_OCTET_STRING;
+        subject_key_id->p = *p;
+        *p += len;
+    }
+
+    return(0);
+}
+
+/*
+ * AuthorityKeyIdentifier ::= SEQUENCE {
+ *        keyIdentifier [0] KeyIdentifier OPTIONAL,
+ *        authorityCertIssuer [1] GeneralNames OPTIONAL,
+ *        authorityCertSerialNumber [2] CertificateSerialNumber OPTIONAL }
+ *
+ *    KeyIdentifier ::= OCTET STRING
+ */
+static int x509_get_authority_key_id(unsigned char** p,
+    unsigned char* end,
+    mbedtls_x509_authority* authority_key_id)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t len = 0u;
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+        MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE)) != 0)
+    {
+        return(ret);
+    }
+
+    if ((ret = mbedtls_asn1_get_tag(p, end, &len,
+        MBEDTLS_ASN1_CONTEXT_SPECIFIC)) != 0)
+    {
+        /* KeyIdentifier is an OPTIONAL field */
+    }
+    else
+    {
+        authority_key_id->keyIdentifier.len = len;
+        authority_key_id->keyIdentifier.p = *p;
+        authority_key_id->keyIdentifier.tag = MBEDTLS_ASN1_OCTET_STRING;
```

Comment:
Hi Gilles!

This finding is not clear for me, could you help me understand it?
A have found "MBEDTLS_ASN1_CONTEXT_SPECIFIC | 0" only once in mbedtls_x509write_crt_set_authority_key_identifier test. Are you talking about this? What comment should I add and where?

I do not understand this sentence:
"an OCTET STRING in DER must be expressed as a primitive format, hence the tag MBEDTLS_ASN1_CONTEXT_SPECIFIC (without CONSTRUCTED) with a content that's the payload of an OCTET STRING"

Sorry, I am not very familiar with this standard. We are just using PEM formats and we are processing the parsed fields, so most of the deep details are misty for me.

Reply:
I'm not an ASN.1 wizard either, and it took me a while to understand why the code is correct. A naive reading of RFC 5280 would seem to imply that the binary data should have the tag 0x04 which indicates an OCTET STRING. In fact, the general rules about ASN.1 notation mean that the tag should be 0x80 because the `[0] … OPTIONAL` indicates that this is a CONTEXT SPECIFIC with a variant tag of 0. But the payload format is the same as OCTET STRING, so it does make sense after parsing.

You presumably understand this since you were able to write a correct implementation. I understand this after investigation. Please explain it in comments for people who will maintain this code.

Reply:
The code containing "CONTEXT_SPECIFIC | 0" in x509write_crt.c was already in the library as I see, so in my part I just simply used OCTET_STRING and it worked. I haven't met this part of the code before your finding. This is as new for me as for you so I would say it was more of luck than deep knowledge.

Where would you prefer the explanation? To "MBEDTLS_ASN1_CONTEXT_SPECIFIC | 0" or to my code where I load OCTET_STRING?

Thank you!

Reply:
Just above the line `… = MBEDTLS_ASN1_OCTET_STRING`, to explain why this doesn't match the `mbedtls_asn1_get_tag( … CONTEXT_SPECIFIC )` call above.

Reply:
Comment is added